### PR TITLE
allow for world names that don't follow the default scheme

### LINKED ISSUE #47: Could not pass event ChunkLoadEvent
Upon world loading I get an error Could not pass event ChunkLoadEvent to RealisticBiomes v0.6.5.0
 and this stacktrace.

http://pastebin.com/dfwfQB18

### PR DESCRIPTION
**Pull request by gipsy-king**
If there are worlds other than one of world, world_nether,
world_the_end, then a hashmap that maps uuids to thir UUID.hashCode() is
used.
UUID.hashCode() may have collisions, but will hopefully suffice for this
scenario.

should close #47

### CODE REVIEW

**Comment by gipsy-king** [comment]
ok to test

**Comment by ttk2** [comment]
so in theory there could be world name collisions on servers with unreasonable numbers of worlds?

**Comment by gipsy-king** [comment]
In theory, yes. If you're thinking sharding, there should only be one world per shard, if I understand bungee correctly.

**Comment by ttk2** [comment]
that's the plan so it should not be a problem for us.

**Comment by ttk2** [comment]
merging. if thats ok with everyoen?

File: src/com/untamedears/realisticbiomes/persist/WorldID.java [hunk 1]
```diff
@@ -19,27 +24,53 @@ public static void init (Plugin plugin) {
...
-		else if (id == netherID)
+		} else if (id == netherID) {
 			return 1;
-		else if (id == endID)
+		} else if (id == endID) {
 			return 2;
-		else
-			return -1;		
+		} else {
+			for (UUID uid: otherWorlds.keySet()) {
```
Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: UUID has a full featured comparable, and you've got it in a hash, don't waste that ln(n) lookup.

<code>
if (otherWorlds.containsKey(uid)) {
    return otherWorlds.get(uid).intValue();
} else {
    RealisticBiomes.doLog(Level.SEVERE, "Cannot get world id for UUID: " + id);
    return -1;
}
</code>

etc.

File: src/com/untamedears/realisticbiomes/persist/WorldID.java [hunk 2]
```diff
@@ -19,27 +24,53 @@ public static void init (Plugin plugin) {
...
 			return overworldID;
-		else if (id == 1)
+		} else if (id == 1) {
 			return netherID;
-		else if (id == 2)
+		} else if (id == 2) {
 			return endID;
-		else
+		} else {
+			for (UUID uid: otherWorlds.keySet()) {
```
Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: Hashmap has another cool feature: ```map.containsValue()``` -- you can fail-fast on lack-of-value. So wrap your log and null return in an ```if (otherWorlds.containsValue(id))``` check.

File: src/com/untamedears/realisticbiomes/persist/WorldID.java [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
 import org.bukkit.World;
 import org.bukkit.plugin.Plugin;
 
+import com.untamedears.realisticbiomes.RealisticBiomes;
+
 public class WorldID {
 	private static UUID overworldID;
 	private static UUID netherID;
 	private static UUID endID;
+	private static HashMap<UUID, Integer> otherWorlds = null;
```
Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: You don't initialize this anywhere, NPE inbound.

File: src/com/untamedears/realisticbiomes/persist/WorldID.java [hunk 4]
```diff
@@ -19,27 +24,53 @@ public static void init (Plugin plugin) {
...
 		World end = plugin.getServer().getWorld("world_the_end");
 		WorldID.endID = end == null ? null : end.getUID();
+		
+		for (World world: plugin.getServer().getWorlds()) {
+			if (!world.getName().equals("world")
+					&& !world.getName().equals("world_nether")
+					&& !world.getName().equals("world_the_end")) {
+				if (otherWorlds == null) {
+					RealisticBiomes.doLog(Level.WARNING, "There are worlds that don't have default names (world, world_nether, world_the_end), experimental feature");
+					otherWorlds = new HashMap<UUID, Integer>();
```
Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: ignore that other comment, I see it now.

Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: It is worth mentioning, however, that your lookup calls don't check if otherWorlds is initialized or not. Would be safer to always initialize -- defensive programming.

File: src/com/untamedears/realisticbiomes/persist/WorldID.java [hunk 5]
```diff
@@ -19,27 +24,53 @@ public static void init (Plugin plugin) {
...
-		else if (id == 1)
+		} else if (id == 1) {
 			return netherID;
-		else if (id == 2)
+		} else if (id == 2) {
 			return endID;
-		else
+		} else {
+			for (UUID uid: otherWorlds.keySet()) {
+				if (otherWorlds.get(uid).intValue() == id) {
```
Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: autoboxing makes the intValue() call unnecessary.

File: src/com/untamedears/realisticbiomes/persist/WorldID.java [hunk 6]
```diff
@@ -19,27 +24,53 @@ public static void init (Plugin plugin) {
...
 			return 1;
-		else if (id == endID)
+		} else if (id == endID) {
 			return 2;
-		else
-			return -1;		
+		} else {
+			for (UUID uid: otherWorlds.keySet()) {
+				if (uid.equals(id)) {
+					return otherWorlds.get(uid).intValue();
```
Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: autoboxing makes the intValue() call unnecessary.

**Comment by ProgrammerDan** [comment]
@ttk2 a few fixes needed before merge.

File: src/com/untamedears/realisticbiomes/persist/WorldID.java (hunk 5 above)
Comment (gipsy-king) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: Right, thanks

File: src/com/untamedears/realisticbiomes/persist/WorldID.java (hunk 2 above)
Comment (gipsy-king) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: damn, I know I have used containsValue that before. thanks!

Comment (gipsy-king) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: I decided to not to wrap in an `if`, since it will do a similar lookup internally and if it matches it will have done two lookups.

**Comment by gipsy-king** [comment]
@ProgrammerDan updated

**Comment by ttk2** [comment]
@ProgrammerDan  @gipsy-king  good now? If I put this online will it require other changes?

**Comment by gipsy-king** [comment]
@ttk2 no it won't require any changes, but I haven't tested this one nor compiled it.

File: src/com/untamedears/realisticbiomes/persist/WorldID.java (hunk 2 above)
Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: Yeah, that's fine. Only way to get prime efficiency would be to leverage additional memory with an inverse map, but that's just getting silly.

File: src/com/untamedears/realisticbiomes/persist/WorldID.java [hunk 7]
```diff
@@ -19,27 +24,50 @@ public static void init (Plugin plugin) {
...
 			return 0;
-		else if (id == netherID)
+		} else if (id == netherID) {
 			return 1;
-		else if (id == endID)
+		} else if (id == endID) {
 			return 2;
-		else
-			return -1;		
+		} else if (otherWorlds.containsKey(id) {
```
Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: Missing closing ```)``` on the conditional

**Comment by ProgrammerDan** [comment]
should be good now @ttk2, I can test compilation tonight unless you just wanna do it live, heh. 

Somewhat pedantic but @gipsy-king it would be best to include a "otherworlds != null &&" before each utilization, since that variable is not always initialized. I doubt it'll pop up too often but no point leaving NPE-honeypots if you can easily avoid them.

**Comment by gipsy-king** [comment]
@ProgrammerDan wouldnt we then get the previous NPE anyway? Also this code won't be used on civcraft.

**Comment by ProgrammerDan** [comment]
Yeah I was figuring to add it so other servers don't run into NPEs -- as to your question, just the lookup functions. You can't know that someone will pass a valid value in -- something might go wrong when they are using the plugin, for instance, or perhaps they are getting clever and using RB as a dependency for a new plugin -- and if the first three if/else blocks fail to match, and you hit the otherworlds test, NPE unless you test for null.

**Comment by gipsy-king** [comment]
@ProgrammerDan I don't know what to do. In my javascript world where only runtime exception exists, I'd throw a descriptive exception. But here, if I check for null, I still have to return something crappy which will throw the NPE as the guy from the issue was getting. If I throw an exception, I have to try/catch at every call point, and this is out of scope for this case IMO, because then I'd have to add a bunch of logic I don't want to test right now.

The log warning is crappy too for sure, but at least next time we can identify what's wrong. I could imagine strange things this isn't prepared for, like a plugin adding worlds dynamcally at runtime and such, and I don't want to deal with that at all.

Can we just go with it? :smiling_imp:

**Comment by ProgrammerDan** [comment]
Well, all this would be solved if folks wrote some decent javadoc to describe expected return values :P. 

You've got the crux of it, ideally these functions should *intentionally* throw an exception if the input cannot be validated as there is no valid output if the input cannot be matched to a world. However, and it's a practical decision, you're going the C route of returning ```-1``` if the UUID can't be found, and ```null``` if the "id" can't be found. So be consistent. Never introduce a state where instead of returning null, you throw NPE. If there exists a valid state where the function returns a value on invalid input, make that your standard return for invalid input unless whatever causes the NPE is completely beyond your ability to predict.

tl;dr Coding best practice in java: If you return a non-error value for some bad inputs, return it for all expected bad inputs, or stop returning null and always intentionally throw an Exception for bad inputs.

Either that or just document it in javadoc at the top of the method, like "If you have non-standard worlds, but you ask for one the method can't find, returns ```null```. If you don't have non-standard worlds, but ask for a non-standard world, throws NPE."

(I'm so wordy, sorry) Just in general, prevent and control every exception scenario, NPE's are horrible, don't let them pass you by if you can prevent it. I've eaten enough of my own crow even just over the past few weeks for not capturing NPE states and preventing.

**Comment by gipsy-king** [comment]
I am enlightened, `push -f` will follow shortly :)

File: src/com/untamedears/realisticbiomes/persist/WorldID.java (hunk 2 above)
Comment (erocs) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: Why not use entrySet() and forgo the slightly more expensive get() call?

Comment (ProgrammerDan) [inline src/com/untamedears/realisticbiomes/persist/WorldID.java]: No reason not to

**Comment by erocs** [comment]
> Also this code won't be used on civcraft.

When sharding is enabled it will be.

> Null null null

It seems that if otherWorlds is initialized statically (and make otherWorlds final) it would never be null under normal circumstances. The Integer value might be null, but returning null is already an error indicator for getMCID() so that's fine.

**Comment by gipsy-king** [comment]
> When sharding is enabled it will be.

It really shouldn't, one server should have no more than one world, not even nether nor the end, otherwise sharding is kinda pointless.

That of course is no excuse for ambiguos code, I'm following Dan's guidelines.

**Comment by ProgrammerDan** [comment]
I think he meant the names of the worlds would be non-standard, hence the code would be active.

**Comment by gipsy-king** [comment]
Oh right. Let's hope no civplugins depend on "default" world names :sweat_smile:

**Comment by gipsy-king** [comment]
@ProgrammerDan see if I got this right, also using entrySet.

**Comment by ProgrammerDan** [comment]
Yeah, looks better.
